stripped/colabs/Birch_stripped.py

```diff
@@ -0,0 +1,732 @@
+  """
+  Funkcja tworzy wykres 3d dla wybranej wartości perplexity.
+  Parametry:
+  - list_of_perplexities - lista wartości perplexity dla których chcemy stworzyć wykresy,
+  - scaled_df - ramka danych z już przeskalowanymi wartościami na podstawie której wykonujemy algorytm t-SNE,
+  - df_with_target - ramka danych w której znajdują się kolumny "pustynia", "step", "lon" i "lat",
+  - czy_step_to_pustynia - wartość logiczna, gdy TRUE to traktujemy step jako pustynię.
+  """
+  tsne = TSNE(n_components = 3, perplexity = perplexity)
+  tsne_results = tsne.fit_transform(scaled_df)
+
+  df_merged_with_anotated = original_df.merge(anotated_df[['lon', 'lat', 'pustynia', 'step', 'pustynia_i_step']], left_on=['lon','lat'], right_on = ['lon','lat'], how='left')
+  df_merged_with_anotated.fillna(2, inplace=True)
+
+  df = pd.DataFrame()
+  df['dim 1'] = tsne_results[:,0]
+  df['dim 2'] = tsne_results[:,1]
+  df['dim 3'] = tsne_results[:,2]
+  if czy_step_to_pustynia:
+    df['Cluster'] = df_merged_with_anotated["pustynia_i_step"]
+  else:
+    df['Cluster'] = df_merged_with_anotated["pustynia"]
+
+  fig = go.Figure(data=[go.Scatter3d(
+    x=df['dim 1'],
+    y=df['dim 2'],
+    z=df['dim 3'],
+    mode="markers",
+    marker=dict(color=df['Cluster'], colorscale=["darkgreen", "orange"]),
+    text = [f"Współrzędne: ({lat}, {lon})" for lat, lon in df_merged_with_anotated[["lat", "lon"]].values],
+    hoverinfo='text'
+  )])
+  fig.update_layout(scene = dict(
+                    xaxis_title='dim 1',
+                    yaxis_title='dim 2',
+                    zaxis_title='dim 3'))
+  fig.update_layout(scene=dict(xaxis_showspikes=False,
+                             yaxis_showspikes=False))
+  fig.show()
+
+
+# In[ ]:
+
+
+def do_kmeans_with_tsne_return_preds(list_of_scaled_df: List, n_components: int=2, perplexity:float=30, learning_rate: Union[float, str]="auto", n_clusters: int=2, random_state: int=123) -> List:
+  warnings.filterwarnings('ignore')
+  list_of_preds = []
+  for scaled_df in list_of_scaled_df:
+    tsne = TSNE(n_components = n_components, perplexity = perplexity, learning_rate=learning_rate)
+    tsne_results = tsne.fit_transform(scaled_df)
+
+    kmeans = KMeans(n_clusters = n_clusters, init = "k-means++", random_state = random_state)
+    kmeans.fit(tsne_results)
+    pred = kmeans.predict(tsne_results)
+    list_of_preds.append(pred)
+  return list_of_preds
+
+
+# In[ ]:
+
+
+hierarchical_cluster_ward = AgglomerativeClustering(n_clusters=2, affinity='euclidean', linkage='ward')
+labels_ward = hierarchical_cluster_ward.fit_predict(X_scaled.loc[:,'Rainf':'SoilM_100_200cm'])
+
+
+# In[ ]:
+
+
+nasa_202307['Cluster'] = birch_result
+
+
+# In[ ]:
+
+
+np.unique(labels_ward, return_counts=True)
+
+
+# In[ ]:
+
+
+colormap_cluster = get_colormap([0, max(nasa_202307['Cluster'].values)], ['darkgreen', "orange"])
+output_notebook()
+show(plot_map(df=nasa_202307, parameter_name='Cluster', colormap=colormap_cluster, title="Predykce", alpha=0.5))
```

Comment:
Literówka w tytule mapy. Nie rozumiem dlaczego "drukujemy" mapę z wynikami birch, skoro linijke wczesniej robiles hierarchical clustering? To pomyłka?